image_proc/src/resize.cpp

```diff
@@ -0,0 +1,179 @@
+*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*********************************************************************/
+#include <mutex>
+#include <cv_bridge/cv_bridge.h>
+#include <rclcpp/rclcpp.hpp>
+#include <rclcpp/qos.hpp>
+#include <sensor_msgs/msg/camera_info.h>
+#include <sensor_msgs/msg/image.h>
+#include <image_transport/image_transport.h>
+
+#include "resize.hpp"
+
+namespace image_proc
+{
+ResizeNode::ResizeNode(const rclcpp::NodeOptions & options) : rclcpp::Node("ResizeNode", options)
+{
+
+   auto parameter_change_cb =
+    [this](std::vector<rclcpp::Parameter> parameters) -> rcl_interfaces::msg::SetParametersResult
+    {
+      auto result = rcl_interfaces::msg::SetParametersResult();
+      result.successful = true;
+      for (auto parameter : parameters) {
+        if (parameter.get_name() == "camera_namespace") {
+          camera_namespace_ = parameter.as_string();
+          RCLCPP_INFO(get_logger(), "camera_namespace: %s ", camera_namespace_.c_str());
+          break;
+        }
+      }
+
+      image_topic_ = camera_namespace_ + "/image";
+      camera_info_topic_ = camera_namespace_ + "/camera_info";
+      connectCb();
+
+      std::lock_guard<std::mutex> lock(connect_mutex_);
+      pub_image_ = image_transport::create_publisher(this, image_topic_ + "/resize");
+      pub_info_ = this->create_publisher<sensor_msgs::msg::CameraInfo>(camera_info_topic_ + "/resize", rclcpp::QoS(1));
+
+      return result;
+    };
+
+  this->declare_parameter("camera_namespace");
+
+  interpolation_ = this->declare_parameter("interpolation", 1);
+  use_scale_ = this->declare_parameter("use_scale", true);
+  scale_height_ = this->declare_parameter("scale_height", 1.0);
+  scale_width_ = this->declare_parameter("scale_width", 1.0);
+  height_ = this->declare_parameter("height", -1);
+  width_ = this->declare_parameter("width", -1);
+
+  rclcpp::Parameter parameter;
+  if (rclcpp::PARAMETER_NOT_SET != this->get_parameter("camera_namespace", parameter)) {
+    std::cout << "Camera space set" << std::endl;
+    parameter_change_cb(this->get_parameters({"camera_namespace"}));
+  }
+
+  this->set_on_parameters_set_callback(parameter_change_cb);
+};
+
+// Handles (un)subscribing when clients (un)subscribe
+void ResizeNode::connectCb()
+{
+  std::lock_guard<std::mutex> lock(connect_mutex_);
+  
+  if (!sub_image_)
+  {
+    std::cout << "subscribe sub_image_" << std::endl;
```

Comment:
Use RCLCPP logging if not a debug article?

Reply:
Yepp I'll just remove it, some debugging artifacts.